runtime/vm/createramclass.cpp

```diff
@@ -3995,23 +4074,406 @@ allocateRAMClassFragmentFromFreeList(RAMClassAllocationRequest *request, J9RAMCl
+	coalesceFreeBlocks(&freeLists->ramClassSmallBlockFreeList);
+	coalesceFreeBlocks(&freeLists->ramClassLargeBlockFreeList);
+}
+
+/**
+ * @brief Allocates pending RAM class fragments from a new segment.
+ *
+ * Computes the required segment size (incl. alignment and, for SUB4G,
+ * space for the lastAllocatedClass pointer). Skips allocation if no
+ * fragments are pending. Uses a fixed increment for SUB4G segments and
+ * the VM default otherwise. Allocates a segment (tagged SUB4G if needed),
+ * carves it into aligned fragments, records each request->address,
+ * and returns gaps to the free lists.
+ *
+ * @param requests Head of the linked list of allocation requests.
+ * @param allocationRequestCount Number of entries in allocationRequests.
+ * @param javaVM J9JavaVM instance.
+ * @param classLoader Class loader performing the allocation.
+ * @param allocationRequests Array of requests, used for cleanup on failure.
+ * @param j9RamClassFreeList RAM class free-list set.
+ * @param ramClassUDATABlockFreelist Head of the UDATA block free list.
+ * @param segmentKind Segment category to allocate under.
+ * @return TRUE if all fragments were allocated; FALSE on failure.
+ *
+ * @note On failure, any allocated fragments are freed and request
+ *       addresses cleared.
+ */
+static BOOLEAN
+allocateRemainingFragments(RAMClassAllocationRequest *requests, UDATA allocationRequestCount, J9JavaVM *javaVM,
+	J9ClassLoader *classLoader, RAMClassAllocationRequest *allocationRequests,
+	J9RAMClassFreeLists *j9RamClassFreeList, UDATA *ramClassUDATABlockFreelist,  SegmentKind segmentKind)
+{
+	UDATA i = 0;
+	if (NULL != requests) {
+		/* Calculate required space in new segment, including maximum alignment padding. */
+		UDATA newSegmentSize = 0;
+		J9MemorySegment *newSegment = NULL;
+		UDATA allocAddress = 0;
+		BOOLEAN isLoadedByAnonClassLoader = (classLoader == javaVM->anonClassLoader);
+		RAMClassAllocationRequest *request = NULL;
+		UDATA fragmentsLeftToAllocate = 0;
+		BOOLEAN isNotLoadedByAnonClassLoader = (classLoader != javaVM->anonClassLoader);
+
+		for (request = requests; NULL != request; request = request->next) {
+			if (!isNotLoadedByAnonClassLoader
+				|| ((NULL == request->address) && (request->segmentKind == segmentKind))
+			) {
+				fragmentsLeftToAllocate++;
+				newSegmentSize += request->fragmentSize + request->alignment;
+			}
+		}
+
+		/* Avoid coalesce/allocating 0-byte segments. */
+		if (0 == fragmentsLeftToAllocate) {
+			return TRUE;
+		}
+
+		/* Add sizeof(UDATA) to hold the "lastAllocatedClass" pointer. */
+		if (SUB4G == segmentKind) {
+			newSegmentSize += sizeof(UDATA);
+		}
+
+		/* Allocate a new segment of the required size. */
+		UDATA classAllocationIncrement = javaVM->ramClassAllocationIncrement;
+		if (isLoadedByAnonClassLoader) {
+			classAllocationIncrement = 0;
+		} else {
+			if (SUB4G == segmentKind) {
+				const UDATA headersPerSegment = 32;
+				classAllocationIncrement = sizeof(J9Class) * headersPerSegment;
```

Comment:
Should alignment be considered here?
```suggestion
                                UDATA mask = J9_REQUIRED_CLASS_ALIGNMENT - 1;
                                UDATA unitSegment = (sizeof(J9Class) + mask) & ~mask;
                                const UDATA headersPerSegment = 32;
                                classAllocationIncrement = unitSegment * headersPerSegment;
```